module_utils/certificate/providers/base.py

```diff
@@ -0,0 +1,291 @@
+
+        return self_info == other_info
+
+
+class CertificateRequestBaseProvider:
+    """Base class for Certificate Request Providers."""
+
+    certificate_proxy_class = CertificateProxy
+
+    __metaclass__ = ABCMeta
+
+    def __init__(self, ansible_module):
+        self.module = ansible_module
+        self.message = ""
+        self.changed = False
+
+        self._existing_certificate = None
+        self._csr = None
+
+    def _run_command(self, *args, **kwargs):
+        """Proxy run_command from Ansible module.
+
+        This is a utility method and does not change the original method
+        in anyway.
+        """
+        return self.module.run_command(*args, **kwargs)
+
+    # TODO: OS dependent
+    def _get_store_location(self, key=False):
+        if key:
+            ext = ".key"
+            subdir = "private"
+        else:
+            ext = ".crt"
+            subdir = "certs"
+
+        name = self.module.params.get("name")
+        if os.path.isabs(name):
+            return name + ext
+
+        base_path = "/etc/pki/tls"
+        return os.path.join(base_path, subdir, (name + ext),)
+
+    @property
+    def existing_certificate(self):
+        """Instance CertificateProxy for the current certificate.
+
+        If no CSR or Certificate are currently available, return None instead.
+        """
+        if self._existing_certificate is None:
+            if not os.path.exists(self.certificate_file_path):
+                self._existing_certificate = False
+            else:
+                csr_pem = self.get_existing_csr_pem_data()
+                cert_pem = self.get_existing_certificate_pem_data()
+
+                self._existing_certificate = self.certificate_proxy_class.load_from_pem(
+                    cert_pem, csr_pem,
+                )
+        return self._existing_certificate
+
+    @property
+    def csr(self):
+        """Instance CertificateProxy for the new certificate request."""
+        if self._csr is None:
+            self._csr = self.certificate_proxy_class.load_from_params(
+                self.module.params,
+            )
+        return self._csr
+
+    @property
+    def certificate_file_path(self):
+        """Path where certificate should be placed."""
+        return self._get_store_location()
+
+    @property
+    def certificate_key_path(self):
+        """Path where key should be placed."""
+        return self._get_store_location(key=True)
+
+    @property
+    def cert_needs_update(self):
+        """Verify if the existing certificate needs to be updated."""
+        return self.existing_certificate != self.csr
+
+    def _exit_success(self):
+        self.module.exit_json(
+            changed=self.changed, msg=self.message,
+        )
+
+    def run(self, check_mode=False):
+        """Entry point for the providers called from the actual Ansible module."""
+        if check_mode:
+            self.message += "(Check mode) "
+
+        self.message = "Certificate "
```

Comment:
This will overwrite the previous (Check mode) if check_mode == True

Reply:
f47b48dab848b02f924458f6a59ac75222c7a944c

Reply:
47b48dab848b02f924458f6a59ac75222c7a944c